Prep work for dual-funding branch (I of IV)

niftynei opened the pull request:
Bunch of small things that came up while working on the dual-funding implementation.

rustyrussell commented on common/amount.c:
```diff
@@ -421,6 +421,10 @@ struct amount_sat amount_tx_fee(u32 fee_per_kw, size_t weight)
+	/* Handle zero case */
+	if (fee_per_kw == 0)
+		return AMOUNT_SAT(0);
+
```
True, but the rest of the code is very cheap, so this seems a weird case to optimize?

http://www2.ing.unipi.it/~a009435/issw/extra/kp_elems_of_pgmng_sty.pdf Page 131.

niftynei commented on tools/test/test_cases:
```diff
@@ -47,6 +47,10 @@ msgdata,test_msg,test_sbt_arrays,subtype_arrays,
+msgtype,test_ifdef,100
+#ifdef TEST_IFDEF
+msgdata,test_ifdef,is_def,u32,
+
```
so there's two problems here. one is that i really wanted #if not #ifdef. i have a fix for this, i need to push it up still.

the parser ends up 'wrapping' the entire message that it's embedded in with the #if tag. 

```example.c
#ifdef TEST_IFDEF
void towire_testifdef(...)
{
   ...
}
#endif /* TEST_IFDEF */
```

as for why it's embedded after the msgtype:  it's marginally easier to assign but changing it to apply to the next parsed msgtype line is probably the right idea. so they'd be declared as 

```
#if TEST_IF
msgtype, ... 
```

Alternatively we could use a different syntax, i.e. just getting rid of the octothorpe. Originally it was patterned after the `#include` tags that we pass through also

```
if TEST_IF
msgtype, ...
```

rustyrussell replied:
But what we really want in some cases is "this field is in the msg, IFF <something>".  But we can bikeshed it later...

niftynei replied:
oh true. this just works on a message level basis ... hmm

niftynei commented:
Addressed all the comments except for modifying how #if declaration works. I think this needs a bit of work to get the exact behavior that we'd like, but I need it as is for later commits in dual funding, so as not to break normal (non EXPERIMENTAL_FEATURES) builds.

niftynei commented on bitcoin/tx.c:
```diff
@@ -57,6 +57,9 @@ int bitcoin_tx_add_input(struct bitcoin_tx *tx, const struct bitcoin_txid *txid,
 
 	/* Now store the input amount if we know it, so we can sign later */
+	if (tal_count(tx->input_amounts) < tx->wtx->num_inputs)
+		tal_resize(tx->input_amounts, tx->wtx->num_inputs);
```
Updated

rustyrussell commented:
Removed erroneous free in last commit (was freed already as child of cmd):

```
    ==51807== Invalid read of size 8
    ==51807==    at 0x1B4505: to_tal_hdr (tal.c:174)
    ==51807==    by 0x1B4EB1: tal_free (tal.c:474)
    ==51807==    by 0x196A33: wallet_withdrawal_broadcast (walletrpc.c:74)
    ==51807==    by 0x115A80: process_sendrawtx (bitcoind.c:449)
    ==51807==    by 0x1150B0: bcli_finished (bitcoind.c:227)
    ==51807==    by 0x1A6A93: destroy_conn (poll.c:244)
    ==51807==    by 0x1A6AB3: destroy_conn_close_fd (poll.c:250)
    ==51807==    by 0x1B46BE: notify (tal.c:235)
    ==51807==    by 0x1B4BAD: del_tree (tal.c:397)
    ==51807==    by 0x1B4F39: tal_free (tal.c:481)
    ==51807==    by 0x1A52B2: io_close (io.c:450)
    ==51807==    by 0x1A71CE: io_loop (poll.c:449)
    ==51807==  Address 0x5af1488 is 24 bytes inside a block of size 120 free'd
    ==51807==    at 0x483997B: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==51807==    by 0x1B4C7B: del_tree (tal.c:416)
    ==51807==    by 0x1B4BFF: del_tree (tal.c:407)
    ==51807==    by 0x1B4F39: tal_free (tal.c:481)
    ==51807==    by 0x12AC8F: command_raw_complete (jsonrpc.c:444)
    ==51807==    by 0x12AD26: command_success (jsonrpc.c:456)
    ==51807==    by 0x1969F1: wallet_withdrawal_broadcast (walletrpc.c:68)
    ==51807==    by 0x115A80: process_sendrawtx (bitcoind.c:449)
    ==51807==    by 0x1150B0: bcli_finished (bitcoind.c:227)
    ==51807==    by 0x1A6A93: destroy_conn (poll.c:244)
    ==51807==    by 0x1A6AB3: destroy_conn_close_fd (poll.c:250)
    ==51807==    by 0x1B46BE: notify (tal.c:235)
    ==51807==  Block was alloc'd at
    ==51807==    at 0x483874F: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==51807==    by 0x1B4728: allocate (tal.c:245)
    ==51807==    by 0x1B4CC1: tal_alloc_ (tal.c:423)
    ==51807==    by 0x196DEC: json_prepare_tx (walletrpc.c:171)
    ==51807==    by 0x19792D: json_txprepare (walletrpc.c:347)
    ==51807==    by 0x12B617: parse_request (jsonrpc.c:650)
    ==51807==    by 0x12BA51: read_json (jsonrpc.c:748)
    ==51807==    by 0x1A4597: next_plan (io.c:59)
    ==51807==    by 0x1A5114: do_plan (io.c:407)
    ==51807==    by 0x1A5152: io_ready (io.c:417)
    ==51807==    by 0x1A71A7: io_loop (poll.c:445)
    ==51807==    by 0x127E14: io_loop_with_timers (io_loop_with_timers.c:24)
```

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/pull/3121/commits/aab57898b0958d59c99552bf94f9fd6620044133
